rules/manager_test.go

```diff
@@ -193,6 +195,171 @@ func TestAlertingRule(t *testing.T) {
+			http_requests{job="app-server", instance="1", group="canary", severity="overwrite-me"}	80 90 100 110 120 130 140
+	`)
+	t.Cleanup(func() {
+		storage.Close()
+	})
+
+	expr, err := parser.ParseExpr(`http_requests{group="canary", job="app-server"} < 100`)
+	require.NoError(t, err)
+
+	newRule := NewAlertingRule(
+		"HTTPRequestRateLow",
+		expr,
+		time.Minute,
+		0,
+		labels.FromStrings("severity", "{{\"c\"}}ritical"),
+		labels.EmptyLabels(), labels.EmptyLabels(), "", false, nil,
+	)
+
+	recordingRule := NewRecordingRule(
+		"HTTPRequestRate",
+		expr,
+		labels.EmptyLabels())
+
+	var ruleGroupActive atomic.Bool
+	ruleGroupActive.Store(true)
+
+	evalRuleGroupFunc := func(ctx context.Context, g *Group, evalTimestamp time.Time, doneCh <-chan struct{}) bool {
+		return ruleGroupActive.Load()
+	}
+
+	var iterations atomic.Int32
+	var shouldMissEval atomic.Bool
+	shouldMissEval.Store(false)
+	evalIterationFunc := func(ctx context.Context, g *Group, evalTimestamp time.Time) {
+		iterations.Add(1)
+		if shouldMissEval.Load() {
+			time.Sleep(2000 * time.Millisecond)
+			DefaultEvalIterationFunc(ctx, g, evalTimestamp)
+			return
+		}
+		modifiedTS := time.Unix(0, 0).Add(time.Duration(iterations.Load()) * time.Second)
+		DefaultEvalIterationFunc(ctx, g, modifiedTS)
+	}
+
+	registry := prometheus.NewRegistry()
+
+	opts := &ManagerOptions{
+		ExternalURL:     nil,
+		QueryFunc:       EngineQueryFunc(testEngine, storage),
+		NotifyFunc:      func(ctx context.Context, expr string, alerts ...*Alert) {},
+		Context:         context.Background(),
+		Appendable:      storage,
+		Queryable:       storage,
+		Logger:          log.NewNopLogger(),
+		OutageTolerance: 30 * time.Minute,
+		ForGracePeriod:  10 * time.Minute,
+		Registerer:      registry,
+		Metrics:         nil,
+	}
+
+	group := NewGroup(GroupOptions{
+		Name:                        "active-passive",
+		Interval:                    time.Second,
+		Rules:                       []Rule{newRule, recordingRule},
+		ShouldRestore:               true,
+		Opts:                        opts,
+		EvalIterationFunc:           evalIterationFunc,
+		ShouldEvaluateRuleGroupFunc: evalRuleGroupFunc,
+	})
+
+	gatherMetrics := func() (map[string]float64, error) {
+		ms, err := registry.Gather()
+		if err != nil {
+			return nil, err
+		}
+		result := make(map[string]float64)
+		for _, m := range ms {
+			for _, mn := range metricNames {
+				if m.GetName() == mn {
+					if m.GetType() == io_prometheus_client.MetricType_COUNTER {
+						for _, k := range m.GetMetric() {
+							result[m.GetName()] = k.Counter.GetValue()
+						}
+					}
+					if m.GetType() == io_prometheus_client.MetricType_GAUGE {
+						for _, k := range m.GetMetric() {
+							result[m.GetName()] = k.Gauge.GetValue()
+						}
+					}
+				}
+			}
+		}
+		return result, nil
+	}
+
+	ctx := context.Background()
+	go func(ctx context.Context) {
+		group.run(ctx)
+	}(ctx)
+
+	time.Sleep(2 * time.Second)
```

Comment:
Since you have control over the `evalIterationFunc`, you can use WaitGroups instead of using sleeps, which can make tests flaky and longer than we need.

Reply:
I changed the test to use channels instead of `time.Sleep`